File: sample/py/13_maze_generation_steps.py
# 13: DFS迷路生成の進行状況をGIF出力するサンプル。

from __future__ import annotations

from time import perf_counter

from py_module.gif_helper import grayscale_palette, save_gif


def capture(grid: list[list[int]], w: int, h: int, scale: int) -> bytes:
    width = w * scale
    height = h * scale
    frame = bytearray(width * height)
    y = 0
    while y < h:
        x = 0
        while x < w:
            v = 255 if grid[y][x] == 0 else 40
            yy = 0
            while yy < scale:
                base = (y * scale + yy) * width + x * scale
                xx = 0
                while xx < scale:
                    frame[base + xx] = v
                    xx += 1
                yy += 1
            x += 1
        y += 1
    return bytes(frame)


def run_13_maze_generation_steps() -> None:
    cell_w = 61
    cell_h = 45
    scale = 4
    out_path = "sample/out/13_maze_generation_steps.gif"

    start = perf_counter()
    grid: list[list[int]] = []
    gy = 0
    while gy < cell_h:
        row: list[int] = []
        gx = 0
        while gx < cell_w:
            row.append(1)
            gx += 1
        grid.append(row)
        gy += 1
    stack: list[tuple[int, int]] = [(1, 1)]
    grid[1][1] = 0

    dirs: list[tuple[int, int]] = [(2, 0), (-2, 0), (0, 2), (0, -2)]
    frames: list[bytes] = []
    step = 0

    while len(stack) > 0:
        x, y = stack[-1]
        candidates: list[tuple[int, int, int, int]] = []
        k = 0
        while k < 4:
            dx, dy = dirs[k]
            nx = x + dx
            ny = y + dy
            if nx >= 1 and nx < cell_w - 1 and ny >= 1 and ny < cell_h - 1 and grid[ny][nx] == 1:
                if dx == 2:
                    candidates.append((nx, ny, x + 1, y))
                elif dx == -2:
                    candidates.append((nx, ny, x - 1, y))
                elif dy == 2:
                    candidates.append((nx, ny, x, y + 1))
                else:
                    candidates.append((nx, ny, x, y - 1))
            k += 1

        if len(candidates) == 0:
            stack.pop()
        else:
            sel = candidates[(x * 17 + y * 29 + len(stack) * 13) % len(candidates)]
            nx, ny, wx, wy = sel
            grid[wy][wx] = 0
            grid[ny][nx] = 0
            stack.append((nx, ny))

        if step % 25 == 0:
            frames.append(capture(grid, cell_w, cell_h, scale))
        step += 1

    frames.append(capture(grid, cell_w, cell_h, scale))
    save_gif(out_path, cell_w * scale, cell_h * scale, frames, grayscale_palette(), delay_cs=4, loop=0)
    elapsed = perf_counter() - start
    print("output:", out_path)
    print("frames:", len(frames))
    print("elapsed_sec:", elapsed)


if __name__ == "__main__":
    run_13_maze_generation_steps()

File: sample/lua/13_maze_generation_steps.lua
dofile((debug.getinfo(1, "S").source:sub(2):match("^(.*[\\/])") or "") .. "py_runtime.lua")

-- from __future__ import annotations as annotations (not yet mapped)
local perf_counter = __pytra_perf_counter
local grayscale_palette = __pytra_grayscale_palette
local save_gif = __pytra_save_gif

-- 13: Sample that outputs DFS maze-generation progress as a GIF.

function capture(grid, w, h, scale)
    width = (w * scale)
    height = (h * scale)
    frame = __pytra_bytearray((width * height))
    for y = 0, h - 1 do
        for x = 0, w - 1 do
            v = (((grid[(((y) < 0) and (#(grid) + (y) + 1) or ((y) + 1))][(((x) < 0) and (#(grid[(((y) < 0) and (#(grid) + (y) + 1) or ((y) + 1))]) + (x) + 1) or ((x) + 1))] == 0)) and (255) or (40))
            for yy = 0, scale - 1 do
                base = ((((y * scale) + yy) * width) + (x * scale))
                for xx = 0, scale - 1 do
                    frame[((((base + xx)) < 0) and (#(frame) + ((base + xx)) + 1) or (((base + xx)) + 1))] = v
                end
            end
        end
    end
    return __pytra_bytes(frame)
end

function run_13_maze_generation_steps()
    -- Increase maze size and render resolution to ensure sufficient runtime.
    cell_w = 89
    cell_h = 67
    scale = 5
    capture_every = 20
    out_path = "sample/out/13_maze_generation_steps.gif"
    
    start = perf_counter()
    local grid = (function() local __lc_out_6 = {}; for __lc_i_5 = 0, (cell_h) - 1, 1 do table.insert(__lc_out_6, __pytra_repeat_seq({ 1 }, cell_w)) end; return __lc_out_6 end)()
    local stack = { { 1, 1 } }
    grid[2][2] = 0
    
    local dirs = { { 2, 0 }, { (-2), 0 }, { 0, 2 }, { 0, (-2) } }
    local frames = {  }
    step = 0
    
    while __pytra_truthy(stack) do
        local __pytra_tuple_8 = stack[(#(stack) + (-1) + 1)]
        x = __pytra_tuple_8[1]
        y = __pytra_tuple_8[2]
        local candidates = {  }
        for k = 0, 4 - 1 do
            local __pytra_tuple_10 = dirs[(((k) < 0) and (#(dirs) + (k) + 1) or ((k) + 1))]
            dx = __pytra_tuple_10[1]
            dy = __pytra_tuple_10[2]
            nx = (x + dx)
            ny = (y + dy)
            if ((nx >= 1) and (nx < (cell_w - 1)) and (ny >= 1) and (ny < (cell_h - 1)) and (grid[(((ny) < 0) and (#(grid) + (ny) + 1) or ((ny) + 1))][(((nx) < 0) and (#(grid[(((ny) < 0) and (#(grid) + (ny) + 1) or ((ny) + 1))]) + (nx) + 1) or ((nx) + 1))] == 1)) then
                if (dx == 2) then
                    table.insert(candidates, { nx, ny, (x + 1), y })
                else
                    if (dx == (-2)) then
                        table.insert(candidates, { nx, ny, (x - 1), y })
                    else
                        if (dy == 2) then
                            table.insert(candidates, { nx, ny, x, (y + 1) })
                        else
                            table.insert(candidates, { nx, ny, x, (y - 1) })
                        end
                    end
                end
            end
        end
        if (#(candidates) == 0) then
            table.remove(stack)
        else
            sel = candidates[(((((((x * 17) + (y * 29)) + (#(stack) * 13)) % #(candidates))) < 0) and (#(candidates) + (((((x * 17) + (y * 29)) + (#(stack) * 13)) % #(candidates))) + 1) or ((((((x * 17) + (y * 29)) + (#(stack) * 13)) % #(candidates))) + 1))]
            local __pytra_tuple_11 = sel
            nx = __pytra_tuple_11[1]
            ny = __pytra_tuple_11[2]
            wx = __pytra_tuple_11[3]
            wy = __pytra_tuple_11[4]
            grid[(((wy) < 0) and (#(grid) + (wy) + 1) or ((wy) + 1))][(((wx) < 0) and (#(grid[(((wy) < 0) and (#(grid) + (wy) + 1) or ((wy) + 1))]) + (wx) + 1) or ((wx) + 1))] = 0
            grid[(((ny) < 0) and (#(grid) + (ny) + 1) or ((ny) + 1))][(((nx) < 0) and (#(grid[(((ny) < 0) and (#(grid) + (ny) + 1) or ((ny) + 1))]) + (nx) + 1) or ((nx) + 1))] = 0
            table.insert(stack, { nx, ny })
        end
        if ((step % capture_every) == 0) then
            table.insert(frames, capture(grid, cell_w, cell_h, scale))
        end
        step = step + 1
    end
    table.insert(frames, capture(grid, cell_w, cell_h, scale))
    save_gif(out_path, (cell_w * scale), (cell_h * scale), frames, grayscale_palette())
    elapsed = (perf_counter() - start)
    __pytra_print("output:", out_path)
    __pytra_print("frames:", #(frames))
    __pytra_print("elapsed_sec:", elapsed)
end


run_13_maze_generation_steps()
